### ISSUE TITLE
Makemigration fails after changes in AbstractRendition

### ISSUE THREAD

**Issue by m1kola** [torchbox/wagtail#2817]
If installation have a custom `Rendition` model (inherited from the `AbstractRendition`), it's not possible to run `makemigrations` command.

In a `master` branch `AbstractRendition.filter` is no longer `ForeignKey`. I'm not sure that we can fix it on the Wagtail's side, but we definitely need to describe steps how to fix it in the related [Upgrade considerations](http://docs.wagtail.io/en/latest/releases/1.6.html#wagtailimages-filter-model-has-been-removed) section.

```python
Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/core/management/__init__.py", line 345, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/core/management/base.py", line 348, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/core/management/base.py", line 399, in execute
    output = self.handle(*args, **options)
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/core/management/commands/makemigrations.py", line 105, in handle
    loader.project_state(),
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/db/migrations/loader.py", line 338, in project_state
    return self.graph.make_state(nodes=nodes, at_end=at_end, real_apps=list(self.unmigrated_apps))
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/db/migrations/graph.py", line 280, in make_state
    project_state = self.nodes[node].mutate_state(project_state, preserve=False)
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/db/migrations/migration.py", line 88, in mutate_state
    operation.state_forwards(self.app_label, new_state)
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/db/migrations/operations/models.py", line 158, in state_forwards
    apps = state.apps
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/utils/functional.py", line 33, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/db/migrations/state.py", line 162, in apps
    return StateApps(self.real_apps, self.models)
  File "/home/vagrant/.virtualenvs/venvname/lib/python3.4/site-packages/django/db/migrations/state.py", line 235, in __init__
    raise ValueError(self._pending_models_error(pending_models))
ValueError: Unhandled pending operations for models:
  wagtailimages.filter (referred to by fields: images.Rendition.filter)
```

**Comment by gasman** [comment]
The error can be prevented by renaming the new non-model `Filter` class to something else (I went with `FilterRunner` - come to think of it, keeping the name `Filter` and moving it to somewhere outside wagtailimages.models is probably nicer) and reinstating the `Filter` model (it only needs to exist - no need to actually reference it anywhere via a foreign key):

    class Filter(models.Model):
        spec = models.CharField(max_length=255, unique=True)

However, the migration generated by makemigrations will just alter the field from a ForeignKey to a CharField without doing the proper data migration.

I think our best bet here is to treat this as a deprecation that happens over several releases, as per our usual deprecation process. In 1.6:

* introduce a nullable `filter_spec` CharField on AbstractRendition
* preserve the `filter` ForeignKey on AbstractRendition and the Filter model, but don't use them
* Add a startup system check that looks for null values in `filter_spec`, and if any exist, abort, telling the user they need to create a data migration
* Provide helper functions for the forward/reverse data migrations (see https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailimages/migrations/0015_remove_filter_model_2.py) so that we can provide an example migration in the release notes that's as short and simple as possible

In 1.8:

* drop the `filter` ForeignKey, the Filter model, and the system check

**Comment by gasman** [comment]
Hmm. Something that complicates the above strategy (of keeping `filter` and `filter_spec` as two columns side by side): when does the`unique_together` constraint get switched from `('image', 'filter', 'focal_point_key')` to `('image', 'filter_spec', 'focal_point_key')`? We can only do this _after_ the data migration has been created and run (because up to that point `filter_spec` will be null everywhere, and the rows won't be unique), which means we can't make the change in models.py (because then that a user installing the new Wagtail release and running `manage.py makemigrations` will end up with that step inside the schema migration).

I think we're going to have to roll this change out even more gradually than anticipated, over several releases. Basically, the need to support custom image models means that we're prevented from doing any "clever" migrations on image models (such as the ones in #2537 / #2755), as that requires us to give deep and detailed instructions to users with custom image models telling them what they should do in place of `./manage.py makemigrations`. I think that's too much of a burden to place on developers - it's already hard enough to convince people to upgrade... :-)

I propose that we set ourselves the following policy: Whenever a change in Wagtail requires developers to create migrations within their own apps on a version upgrade, the migrations will consist of, at most:
 * the schema migration auto-generated by `./manage.py makemigrations`
 * a data migration to be created manually by the developer and applied on top of the schema migration; we will use system checks to alert the developer when this has not been done, and provide helper functions to make the migration code as straightforward as possible.

Since it's probably too late to rethink this in time for 1.6, I suggest that we roll back #2537 / #2755 and begin the process in 1.7. The above policy means that we're limited in how much we can do in each release - I expect the timeline will be something along the lines of:

* 1.7: Add a nullable `filter_spec` column which gets populated alongside the existing `filter` field when creating new Rendition records, but doesn't get used by anything yet; add a data migration to populate this for existing records
* 1.8: Convert Filter to a non-model-based class, as per #2537, and start using the `filter_spec` column; switch the `unique_together` constraint to that column, and drop the `filter` column and `Filter` table

**Comment by gasman** [comment]
Solved by reverting this change, in #2880. Have opened #2881 to deal with re-implementing the change in a way that doesn't trigger this problem.
